File: sdk/nodejs/src/readApi.mjs
/**
 * Read-side helpers for the AI Audit Ledger API.
 *
 * The ingest path is fire-and-forget by design and needs no read setup.
 * This module covers the small read-side calls that complement it: today
 * verifyCompleteness; future verifyDecision and listDecisions if needed.
 *
 * Auth uses a read key (separate namespace from the write key). Read keys
 * cannot write; write keys cannot read.
 */

import { AuditLedgerError } from './index.mjs';

/**
 * @typedef {object} CompletenessResult
 * @property {string}              tenant_id
 * @property {{from: number, to: number}} range
 * @property {number}              expected_count
 * @property {number}              found_count
 * @property {number[]}            missing
 * @property {string}              note
 */

/**
 * Detect deleted or omitted audit records for the caller's tenant.
 *
 * The ledger compares its per-tenant monotonic sequence counter against
 * the records actually present in storage and returns any sequence numbers
 * that are missing. Each gap is a record that was deleted, lost during SQS
 * redelivery, or never stored. The processor logs burned_sequence entries
 * for the redelivery case so operators can distinguish deliberate deletion
 * from observable infrastructure noise.
 *
 * @param {object}  opts
 * @param {string}  opts.apiUrl     — Base API URL (no trailing /audit segment).
 * @param {string}  opts.readKey    — Tenant read key, sent as x-api-key.
 * @param {number}  [opts.from]     — Inclusive lower bound on sequence_no.
 * @param {number}  [opts.to]       — Inclusive upper bound on sequence_no.
 * @param {string}  [opts.tenantId] — Required only with the admin read key.
 * @param {number}  [opts.timeoutMs=10000]
 * @param {number}  [opts.retries=3]
 * @param {typeof globalThis.fetch} [opts.fetchImpl]
 * @returns {Promise<CompletenessResult>}
 */
export async function verifyCompleteness({
  apiUrl,
  readKey,
  from,
  to,
  tenantId,
  timeoutMs = 10_000,
  retries = 3,
  fetchImpl = globalThis.fetch,
}) {
  const base = apiUrl.replace(/\/+$/, '');
  const params = new URLSearchParams();
  if (typeof from === 'number') params.set('from', String(from));
  if (typeof to   === 'number') params.set('to',   String(to));
  if (tenantId) params.set('tenant_id', tenantId);
  const url = `${base}/audit/verify-completeness${params.toString() ? `?${params}` : ''}`;

  let lastErr;
  for (let attempt = 0; attempt < retries; attempt++) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const res = await fetchImpl(url, {
        method: 'GET',
        headers: { Accept: 'application/json', 'x-api-key': readKey },
        signal: controller.signal,
      });
      clearTimeout(timer);
      if (res.ok) return /** @type {CompletenessResult} */ (await res.json());
      if (res.status < 500) {
        const body = await res.text().catch(() => '');
        throw new AuditLedgerError(
          `verify_completeness failed: HTTP ${res.status} ${body}`,
          res.status,
        );
      }
      lastErr = new AuditLedgerError(`HTTP ${res.status}`, res.status);
    } catch (err) {
      clearTimeout(timer);
      // Only swallow into retry on network/timeout errors, not on our own
      // AuditLedgerError thrown from a 4xx body — those are terminal.
      if (err instanceof AuditLedgerError && err.status && err.status < 500) {
        throw err;
      }
      lastErr = err;
    }
    if (attempt < retries - 1) {
      const baseMs = 200;
      const jitter = Math.random() * baseMs;
      await new Promise((r) => setTimeout(r, baseMs * 2 ** attempt + jitter));
    }
  }
  throw lastErr;
}

File: sdk/python/ai_audit_ledger/read_api.py
"""
Read-side helpers for the AI Audit Ledger API.

The ingest client is intentionally write-only: fire-and-forget logging
should not need any read setup. This module provides the read-side calls
that complement it:

  - verify_decision     tamper-evidence check for one specific record
  - verify_completeness proof that no records have been deleted
  - list_decisions      browse recent records for the calling tenant

Each function has both an async form (for use inside an asyncio loop) and
a sync wrapper (for scripts and notebooks).

Auth uses a read key (separate namespace from the write key). Read keys
cannot write; write keys cannot read.
"""

from __future__ import annotations

import asyncio
from typing import Any, Optional, TypedDict
from urllib.parse import quote

import aiohttp

from ai_audit_ledger.client import AuditLedgerError, _with_retry


# ── response shapes ──────────────────────────────────────────────────────────

class DecisionRecord(TypedDict, total=False):
    """Shape of a single audit record as returned by the read API."""
    event_id:           str
    timestamp:          str
    tenant_id:          str
    model_version:      str
    system_prompt_hash: str
    input_data_hash:    str
    ai_decision_output: dict[str, Any]
    human_in_loop:      bool
    sequence_no:        int  # present from v0.3 onward


class TamperCheckResult(TypedDict, total=False):
    """Response from GET /audit/events/{event_id}/history."""
    event_id:           str
    integrity_verified: bool
    integrity_note:     str
    current_record:     DecisionRecord  # the DynamoDB copy
    archived_record:    DecisionRecord  # the S3 Object Lock copy


class CompletenessResult(TypedDict, total=False):
    """Response from GET /audit/verify-completeness."""
    tenant_id:      str
    range:          dict[str, int]
    expected_count: int
    found_count:    int
    missing:        list[int]
    note:           str


class ListDecisionsResult(TypedDict, total=False):
    """Response from GET /audit/logs."""
    items:     list[DecisionRecord]
    count:     int
    tenant_id: str  # omitted for admin callers (cross-tenant)


# ── internal: GET with retry, shared by all three read helpers ───────────────

async def _get_json_with_retry(
    *,
    url: str,
    read_key: str,
    endpoint_name: str,
    params: Optional[dict[str, Any]] = None,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> Any:
    async def attempt() -> aiohttp.ClientResponse:
        timeout = aiohttp.ClientTimeout(total=timeout_s)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            resp = await session.get(
                url,
                params=params,
                headers={"Accept": "application/json", "x-api-key": read_key},
            )
            await resp.read()
            return resp

    resp = await _with_retry(attempt, attempts=retries)
    if resp.status != 200:
        body = (await resp.text()) if not resp.content.is_eof() else ""
        raise AuditLedgerError(
            f"{endpoint_name} failed: HTTP {resp.status} {body}",
            status=resp.status,
        )
    return await resp.json()


# ── verify_decision ──────────────────────────────────────────────────────────

async def verify_decision(
    *,
    api_url: str,
    read_key: str,
    event_id: str,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> TamperCheckResult:
    """
    Tamper-evidence check for one specific recorded decision.

    The ledger fetches both the DynamoDB copy (queryable index) and the S3
    Object Lock copy (immutable archive) of the requested event and compares
    them with stable JSON serialisation. integrity_verified is True when they
    match exactly. A mismatch flips it to False with a warning in integrity_note.

    Args:
        api_url:   Base API URL (no trailing /audit segment).
        read_key:  Tenant read key, sent as x-api-key.
        event_id:  UUID v4 of the event to verify.
        timeout_s: Per-attempt HTTP timeout in seconds.
        retries:   Total attempts (1 = no retry).

    Returns:
        TamperCheckResult dict.

    Raises:
        AuditLedgerError: if the API returns a non-200 status or is unreachable
            after all retry attempts.
        ValueError: if event_id is empty.
    """
    if not event_id:
        raise ValueError("event_id is required for verify_decision")
    base = api_url.rstrip("/")
    url = f"{base}/audit/events/{quote(event_id, safe='')}/history"
    return await _get_json_with_retry(
        url=url,
        read_key=read_key,
        endpoint_name="verify_decision",
        timeout_s=timeout_s,
        retries=retries,
    )


def verify_decision_sync(
    *,
    api_url: str,
    read_key: str,
    event_id: str,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> TamperCheckResult:
    """
    Synchronous wrapper around verify_decision for scripts and notebooks.
    Do not call from inside an async context: use the async version directly.
    """
    return asyncio.run(verify_decision(
        api_url=api_url,
        read_key=read_key,
        event_id=event_id,
        timeout_s=timeout_s,
        retries=retries,
    ))


# ── verify_completeness ──────────────────────────────────────────────────────

async def verify_completeness(
    *,
    api_url: str,
    read_key: str,
    from_seq: Optional[int] = None,
    to_seq: Optional[int] = None,
    tenant_id: Optional[str] = None,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> CompletenessResult:
    """
    Detect deleted or omitted audit records for the caller's tenant.

    The ledger compares its per-tenant monotonic sequence counter against
    the records actually present in storage and returns any sequence
    numbers that are missing. Each gap is a record that was either deleted,
    lost during SQS redelivery, or never stored. The processor logs
    burned_sequence entries for the redelivery case so operators can
    distinguish deliberate deletion from observable infrastructure noise.

    Args:
        api_url:    Base API URL (no trailing /audit segment).
        read_key:   Tenant read key, sent as x-api-key.
        from_seq:   Optional inclusive lower bound. Defaults to 1.
        to_seq:     Optional inclusive upper bound. Defaults to current counter.
        tenant_id:  Required only when calling with the admin read key.
        timeout_s:  Per-attempt HTTP timeout in seconds.
        retries:    Total attempts (1 = no retry).

    Returns:
        CompletenessResult dict.

    Raises:
        AuditLedgerError: if the API returns a non-200 status or is unreachable
            after all retry attempts.
    """
    base = api_url.rstrip("/")
    params: dict[str, Any] = {}
    if from_seq is not None: params["from"]      = from_seq
    if to_seq   is not None: params["to"]        = to_seq
    if tenant_id:            params["tenant_id"] = tenant_id
    return await _get_json_with_retry(
        url=f"{base}/audit/verify-completeness",
        read_key=read_key,
        endpoint_name="verify_completeness",
        params=params,
        timeout_s=timeout_s,
        retries=retries,
    )


def verify_completeness_sync(
    *,
    api_url: str,
    read_key: str,
    from_seq: Optional[int] = None,
    to_seq: Optional[int] = None,
    tenant_id: Optional[str] = None,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> CompletenessResult:
    """
    Synchronous wrapper around verify_completeness for scripts and notebooks.
    Do not call from inside an async context: use the async version directly.
    """
    return asyncio.run(verify_completeness(
        api_url=api_url,
        read_key=read_key,
        from_seq=from_seq,
        to_seq=to_seq,
        tenant_id=tenant_id,
        timeout_s=timeout_s,
        retries=retries,
    ))


# ── list_decisions ───────────────────────────────────────────────────────────

async def list_decisions(
    *,
    api_url: str,
    read_key: str,
    from_ts: Optional[str] = None,
    to_ts: Optional[str] = None,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> ListDecisionsResult:
    """
    Browse recent audit records for the caller's tenant.

    Optional from_ts and to_ts filter to a date range (ISO 8601 strings,
    compared lexicographically against the sort key which encodes timestamp).
    Tenant-scoped automatically by the read key; an admin key returns records
    across all tenants.

    Args:
        api_url:   Base API URL (no trailing /audit segment).
        read_key:  Tenant read key, sent as x-api-key.
        from_ts:   Optional inclusive lower bound on timestamp (ISO 8601).
        to_ts:     Optional inclusive upper bound on timestamp (ISO 8601).
        timeout_s: Per-attempt HTTP timeout in seconds.
        retries:   Total attempts (1 = no retry).

    Returns:
        ListDecisionsResult dict with items, count, and (for tenant callers)
        tenant_id.

    Raises:
        AuditLedgerError: if the API returns a non-200 status or is unreachable
            after all retry attempts.
    """
    base = api_url.rstrip("/")
    params: dict[str, Any] = {}
    if from_ts: params["from"] = from_ts
    if to_ts:   params["to"]   = to_ts
    return await _get_json_with_retry(
        url=f"{base}/audit/logs",
        read_key=read_key,
        endpoint_name="list_decisions",
        params=params,
        timeout_s=timeout_s,
        retries=retries,
    )


def list_decisions_sync(
    *,
    api_url: str,
    read_key: str,
    from_ts: Optional[str] = None,
    to_ts: Optional[str] = None,
    timeout_s: float = 10.0,
    retries: int = 3,
) -> ListDecisionsResult:
    """
    Synchronous wrapper around list_decisions for scripts and notebooks.
    Do not call from inside an async context: use the async version directly.
    """
    return asyncio.run(list_decisions(
        api_url=api_url,
        read_key=read_key,
        from_ts=from_ts,
        to_ts=to_ts,
        timeout_s=timeout_s,
        retries=retries,
    ))
